### PR TITLE
IDSEQ-947 Detect duplicate columns during metadata CSV validation.

### PR DESCRIPTION
**Pull request by MarkAZhang**
Fix issue where user can submit duplicate columns, which causes validation and upload errors.

Duplicate columns are now detected.

Screen Shot 2019-04-29 at 4 04 17 PM

Verification:
* Updated tests and made sure they passed. Including for custom fields (an edge case)
* Verified in the UI that duplicate columns throw error.
* Verified that well-formatted CSVs can be uploaded without error.

### CODE REVIEW

File: app/helpers/metadata_helper.rb [hunk 1]
```diff
@@ -143,6 +143,47 @@ def validate_metadata_csv_for_samples(samples, metadata, new_samples = false)
       return { "errors" => errors, "warnings" => [] }
     end
 
+    existing_fields = MetadataField.where(is_core: 1)
+
+    if samples[0].project
+      existing_fields = (samples[0].project.metadata_fields | MetadataField.where(is_core: 1))
```
[the remark this answers is not in the document]
Comment (MarkAZhang) [reply app/helpers/metadata_helper.rb]: Done.

File: app/helpers/metadata_helper.rb [hunk 2]
```diff
@@ -143,6 +143,47 @@ def validate_metadata_csv_for_samples(samples, metadata, new_samples = false)
...
+    end
+
+    column_to_index = {}
```
[the remark this answers is not in the document]
Comment (MarkAZhang) [reply app/helpers/metadata_helper.rb]: Great point, thanks!

File: app/helpers/metadata_helper.rb [hunk 3]
```diff
@@ -143,6 +143,47 @@ def validate_metadata_csv_for_samples(samples, metadata, new_samples = false)
...
+    has_duplicate_columns = false
+
+    # Prevent duplicate columns.
+    metadata["headers"].each_with_index do |col, index|
+      unless column_to_index[col].nil?
```
[the remark this answers is not in the document]
Comment (MarkAZhang) [reply app/helpers/metadata_helper.rb]: Ah, it's because I didn't know about `present`. You're right, that is way better.

File: app/helpers/metadata_helper.rb [hunk 4]
```diff
@@ -143,6 +143,47 @@ def validate_metadata_csv_for_samples(samples, metadata, new_samples = false)
...
+
+      # Account for both the name and the display_name, as both are accepted.
+      if ["sample_name", "Sample Name"].include?(col)
+        column_to_index["sample_name"] = index
+        column_to_index["Sample Name"] = index
+      elsif ["host_genome", "Host Genome"].include?(col)
+        column_to_index["host_genome"] = index
+        column_to_index["Host Genome"] = index
+      else
+        matching_field = existing_fields.select { |field| field.name == col || field.display_name == col }
```
[the remark this answers is not in the document]
Comment (MarkAZhang) [reply app/helpers/metadata_helper.rb]: Unfortunately, sample name and host genome are special fields that are not stored in MetadataField like everything else, so we have to make an exception for them.
